nutils/sample.py

```diff
@@ -595,6 +595,103 @@ def __call__(self, func: function.IntoArray) -> function.Array:
         return numpy.concatenate([self._sample1(func), self._sample2(func)])
 
 
+def _simplex_strip(strip):
+    # Helper function that creates simplices for an extruded simplex, with
+    # vertices arranged in a [2,n] shape (prepended with an arbitrary number of
+    # axes). The Strategy is to create the first simplex from the first vertex
+    # in layer 1 and all vertices from layer 2, the second from all but the
+    # first of layer 2 and the first two of layer 1, and so on until the last
+    # simplex consists of all vertices of layer 1 and the last of layer 2.
+
+    assert strip.dtype == int
+    *shape, m, n = strip.shape
+    assert m == 2
+    flat = strip.reshape((*shape, 2*n)) # ravel last two axes, reallocates if necessary
+    *strides, s = flat.strides
+    return numpy.ndarray(buffer=flat, dtype=int, shape=(*shape, n, n+1), strides=(*strides, s, s))
+
+
+def _mul_tri(tri1, tri2):
+    # Helper function that computes the tri1 x tri2 product. The indices should
+    # be pre-multiplied with the appropriate strides.
+
+    if tri2 is None: # multiplication with 'empty' right hand side
+        return tri1
+
+    ndims1 = tri1.shape[1] - 1
+    ndims2 = tri2.shape[1] - 1
+
+    tri_outer = tri1[:,None,:,None] + tri2[None,:,None,:]
+    if ndims1 == 0: # Left multiplication by a 0D sample
+        # Multiply the left 0D tri by the right tri and maintain the latter's
+        # triangulation.
+        tri = tri_outer.reshape(-1, ndims2+1)
+    elif ndims2 == 0: # Right multiplication by a 0D sample
+        # Multiply the left tri by the right 0D tri and maintain the former's
+        # triangulation.
+        tri = tri_outer.reshape(-1, ndims1+1)
+    elif ndims1 == 1: # Left multiplication by a 1D sample
+        # Multiply the left 1D tri by the right tri and triangulate using a
+        # simplex strip.
+        tri = _simplex_strip(tri_outer).reshape(-1, ndims2+2)
+    elif ndims2 == 1: # Right multiplication by a 1D sample
+        # Multiply the left 1D tri by the right tri and triangulate using a
+        # simplex strip.
+        tri = _simplex_strip(tri_outer.swapaxes(-2,-1)).reshape(-1, ndims1+2)
+    else:
+        raise NotImplementedError(f'tri not supported for {ndims1}D x {ndims2}D multiplication')
```

Comment:
What do you think of swapping `tri1` and `tri2` after singleton expansion of the first axis if `tri1.shape[-1] > tri2.shape[-1]` and dropping the second and the fourth branch? Same for the hull.

Reply:
Done - and we can even do it before singleton expansion of the first axis as we don't make any promises about the order of the returned triangles.